### PR TITLE
Mega-nix PR

### LINKED ISSUE #152: Nix support for hasktorch-0.2.0.0
From #143:

> ffi-experimental is in master, and it doesn't actively support a nix workflow. [pytorch-world](https://github.com/stites/pytorch-world) (which I've renamed from "libtorch-nix") will be the best place to go to iterate on nix-support. It's also where I've been working on the [pytorch-1.1.0 derivation](https://github.com/NixOS/nixpkgs/pull/65041) -- which works without FBGEMM, and might turn into a 1.1.1 derivation.
>
> If you'd like to work on hasktorch with nix, please help out on that repo (which is properly pinned, following lorri's style). https://github.com/stites/pytorch-world/issues/2 blocks a cmake-less libtorch (which is nessecary for hasktorch), @tscholak has also been helping test that repo on darwin.
>
> If someone starts working on nix support and needs a pin, please pull in pytorch-world as a submodule in a `nix` folder (so `<hasktorch_root>/nix`) and use the pin located at `<hasktorch_root>/nix/pin`.
>
> Refer to https://github.com/stites/probtorch/tree/nix as an example.

This supercedes #103 given all of the latest changes to hasktorch.

### PR DESCRIPTION
**Pull request by stites**
Closes #152 

This PR merges #164, followed by #165, performing some git magic to clean up the merge commits in #165 (and removes my committer dates from rebasing since there were no rebase conflicts). It maintains author commit dates as expected -- a tree visualization would look like the following. Any final clean-up will be amended in the last commit https://github.com/hasktorch/hasktorch/pull/175/commits/2a0f763a9e4f307b6f83e756243e819935ea8fe8. The only difference that I can tell between these two PRs is the removal of `<root>/nixpkgs.json`, which seemed unnecessary.

stites/mega-nix: https://github.com/hasktorch/hasktorch/commit/d5f61710ec6e2c1a01e0b762f272435c79ec8c32#diff-71e6301422957944b76068e5ab438283
tscholak/nix: https://github.com/hasktorch/hasktorch/pull/165/files#diff-71e6301422957944b76068e5ab438283

@tscholak, I can't add you as a reviewer (I believe because you are a participant) -- if you could give a +1 on this change, I think it is ready to be merged in. Please let me know if I am missing something.

@junjihashimoto, @austinvhuang, @tscholak what do you think about this extra [hasktorch/.hie-wrapper.sh](https://github.com/hasktorch/hasktorch/pull/175/files#diff-21f98cfb62b6d4875ce553a325fc0dc9) shell script? I like the idea of packaging IDE tools, but I think it might be better to move this elsewhere. 

Because this PR is blocking a few others (#168, #169, #170) I would opt that this PR gets merged today regardless of the status of this hie script. This script can be moved/renamed/removed later without blocking.

This PR is expected to fail CI, I believe.

---

git log to understand how commits are ordered:

```
* d5f61710 - (HEAD -> mega-nix, origin/mega-nix) clean up mega-nix (2019-09-11 @ 4 minutes ago) <Sam Stites>
*   f6e17058 - Merge branch 'tscholak-nix' into mega-nix (2019-09-11 @ 4 minutes ago) <Sam Stites>
|\
| * cd135137 - (tscholak-nix) pin nix to nixpkg channel (2019-09-05 @ 6 days ago) <Torsten Scholak>
| * 21906ef7 - add all relevant projects to the repo's default.nix (2019-09-05 @ 6 days ago) <Torsten Scholak>
| * fc6c4648 - switch to libtorch from pytorch-world and make libtorch-ffi build on darwin (2019-09-05 @ 6 days ago) <Torsten Scholak>
| * acc62e03 - pin nixpkgs to commit for which pytorch is built with doCheck=false (2019-09-04 @ 7 days ago) <Torsten Scholak>
| * bcd1872c - add shellHook to resolve Intel MKL FATAL ERROR: Cannot load libmkl_avx.so or libmkl_def.so (2019-09-03 @ 8 days ago) <Torsten Scholak>
| * 894a41ab - add HIE wrapper shell script for vscode haskell extension (2019-09-02 @ 9 days ago) <Torsten Scholak>
| * 74ae0043 - link iomp5 from mkl (2019-09-02 @ 9 days ago) <Torsten Scholak>
| * 2f57a195 - don't fail on all errors for examples (2019-09-01 @ 10 days ago) <Torsten Scholak>
| * 2a31c770 - add nix derivation for examples (2019-09-01 @ 10 days ago) <Torsten Scholak>
| * 67791a06 - add default and release nix files (2019-09-01 @ 10 days ago) <Torsten Scholak>
| * b3dc5399 - add nix stubs for hasktorch (2019-09-01 @ 10 days ago) <Torsten Scholak>
| * 75f3f145 - make libtorch-ffi build (2019-09-01 @ 10 days ago) <Torsten Scholak>
| * 6a518fec - use pytorch from nixpkgs instead of libtorch from pytorch-world for now (2019-08-31 @ 11 days ago) <Torsten Scholak>
* |   321574d9 - (mega-nix-1) Merge remote-tracking branch 'wavewave/nix-package' into mega-nix (2019-09-11 @ 82 minutes ago) <Sam Stites>
|\ \
| * | 4019d420 - (wavewave/nix-package) hasktorch is now packaged up. (2019-08-31 @ 11 days ago) <Ian-Woo Kim>
| * | d022b032 - libtorch-ffi is nix-packaged! (2019-08-31 @ 11 days ago) <Ian-Woo Kim>
| * | 33011899 - user-shell with inline-c and inline-c-cpp. (2019-08-31 @ 11 days ago) <Ian-Woo Kim>
| * | 8d6d441d - (wavewave/nix) enable mkl. use dev attribute. (2019-08-31 @ 11 days ago) <Ian-Woo Kim>
| * | d57bffff - remove unnecessary ABI option (2019-08-31 @ 11 days ago) <Ian-Woo Kim>
| * | bf60cb98 - hasktorch works with setting up C++11 ABI flag correctly! (2019-08-31 @ 11 days ago) <Ian-Woo Kim>
| * | 8c01de91 - library_path. (2019-08-31 @ 11 days ago) <Ian-Woo Kim>
| * | ae8c3fb1 - api/include path. temporarily disable mkl (2019-08-31 @ 11 days ago) <Ian-Woo Kim>
| * | bbcaee8d - pin nixpkgs revision. (2019-08-31 @ 11 days ago) <Ian-Woo Kim>
| * | 5cc7d0cb - pytorch 1.2! (2019-08-31 @ 11 days ago) <Ian-Woo Kim>
| * | 6bdb7296 - add more haskell deps and mkl. (2019-08-30 @ 11 days ago) <Ian-Woo Kim>
| * | b8022456 - introduce shell.nix for inline-c (2019-08-30 @ 11 days ago) <Ian-Woo Kim>
| | * 44325bd3 - (nix) add initial nix code (2019-08-30 @ 12 days ago) <Torsten Scholak>
| | |
 ...
```

### CODE REVIEW

**Comment by tscholak** [comment]
I tried it just now on Darwin and it works as expected! Great work, thanks @stites :+1:

**Comment by wavewave** [comment]
Great work! Thanks :+1:  @stites

**Comment by stites** [comment]
I originally assumed that CI failed for reasons relating to nix, but it looks like a recently broken OSX build. Can someone explain what is happening?

**Comment by tscholak** [comment]
@stites 

> I originally assumed that CI failed for reasons relating to nix, but it looks like a recently broken OSX build. Can someone explain what is happening?

https://github.com/hasktorch/hasktorch/blob/2a0f763a9e4f307b6f83e756243e819935ea8fe8/libtorch-ffi/libtorch-ffi.cabal has changed, so maybe these build errors can be attributed to different (c-)flags.

```
    /Users/distiller/project/libtorch-ffi//Users/distiller/project/deps/libtorch/include/c10/core/DeviceType.h:15:12: error:
         error: ISO C++ forbids forward references to 'enum' types
       |
    15 | enum class DeviceType : int16_t {
       |            ^
    enum class DeviceType : int16_t {
```
looks like the compiler is more pedantic now for some reason

**Comment by tscholak** [comment]
@stites note also that https://circleci.com/gh/hasktorch/hasktorch/2045?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link succeeds

**Comment by tscholak** [comment]
@stites looks like this change, https://github.com/hasktorch/hasktorch/pull/168/commits/07383b37a1092e8e0b2836827060dd221391598e, fixed the stack-OS X problems in #168.
But I think this breaks nix on Darwin

**Comment by stites** [comment]
I've cherry-picked that commit to fix CI. This way we maintain compatibility with all non-nix users (the majority) and the darwin-nix workflow will be broken again (heads-up @idontgetoutmuch). This should be the fastest way to get things merged in.

**Comment by austinvhuang** [comment]
Thanks very much @wavewave @tscholak @stites!

I can't add all that much too the review, but with regard to HIE, my bias would be to have it available for those who find it useful. We can always remove later if people find it's not working or getting in the way.

Nice work!

File: libtorch-ffi/libtorch-ffi.cabal [hunk 1]
```diff
@@ -63,11 +63,18 @@ library
...
+  extra-libraries:     c++
+                     , c10
+                     , iomp5
+                     -- , mklml
+                     , torch
+ else
+  extra-libraries:     stdc++
+                     , c10
+                     , iomp5
+                     -- , mklml
```
Comment (junjihashimoto) [inline libtorch-ffi/libtorch-ffi.cabal]: CI  but I don't know if we can delete it,
I'll add a test later to think about what to do.

Comment (junjihashimoto) [reply libtorch-ffi/libtorch-ffi.cabal]: I can not find any dependency of libmklml.os.
So I think it is acceptable.
https://gist.github.com/junjihashimoto/f66a1e34ae8770a35573555e84c0bb2d

Comment (junjihashimoto) [reply libtorch-ffi/libtorch-ffi.cabal]: macos' library-dependency is different from linux's one.
```
bash-3.2$ otool -L libtorch.dylib
libtorch.dylib:
        @rpath/libtorch.dylib (compatibility version 0.0.0, current version 0.0.0)
        /System/Library/Frameworks/Accelerate.framework/Versions/A/Accelerate (compatibility version 1.0.0, current version 4.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
        @rpath/libmklml.dylib (compatibility version 0.0.0, current version 0.0.0)
        @rpath/libiomp5.dylib (compatibility version 5.0.0, current version 5.0.0)
        @rpath/libc10.dylib (compatibility version 0.0.0, current version 0.0.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
bash-3.2$ otool -L libc10.dylib
libc10.dylib:
        @rpath/libc10.dylib (compatibility version 0.0.0, current version 0.0.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
```

Comment (junjihashimoto) [reply libtorch-ffi/libtorch-ffi.cabal]: macos's library needs both libiomp5 and libmklml which linux does not need.

Comment (tscholak) [reply libtorch-ffi/libtorch-ffi.cabal]: great findings, thanks for looking into this @junjihashimoto! when I rewrote this cabal file, I was very conservative and kept any dependency I could fulfil with nix. It’s great that we can get rid of some!

**Pull request merged**